modules/tinymce/src/themes/silver/main/ts/ui/button/MenuButton.ts

```diff
@@ -24,9 +24,13 @@ interface StoredMenuButton extends Omit<Dialog.DialogFooterMenuButton, 'items'>
   readonly items: StoredMenuItem[];
 }
 
-const getMenuButtonApi = (component: AlloyComponent): Toolbar.ToolbarMenuButtonInstanceApi => ({
+const getMenuButtonApi = (spec: MenuButtonSpec, backstage: UiFactoryBackstage) => (component: AlloyComponent): Toolbar.ToolbarMenuButtonInstanceApi => ({
   isEnabled: () => !Disabling.isDisabled(component),
-  setEnabled: (state: boolean) => Disabling.set(component, !state),
+  setEnabled: (state: boolean) => {
+    if (!backstage.shared.providers.isDisabled() || spec.readonly && backstage.shared.providers.isReadOnlyEnableUi()) {
+      Disabling.set(component, !state, true);
```

Comment:
Don't we want to store this no matter what the UI state is? Otherwise someone could set their enabled state while the UI is readonly, and when it is no longer readonly the change they made doesn't apply.

We should _always_ store the developer intention in the disabled state, and only adjust it while rendering based on the UI mode.

- [ ] This applies to all other UI components as well

Reply:
It's required this scenario 😞

onSetup callback with setEnabled(true), but without `allowedModes` in overflow toolbar
```
        ed.ui.registry.addButton('t2', {
          icon: 'italic',
          text: 'Test Menu Item 2',
          onAction: Fun.noop,
          onSetup: (api) => {
            api.setEnabled(true);
            return Fun.noop;
          }
        });
```

The onSetup callback is executed after the buttons are attached in the overflow toolbar

Reply:
Yes, but my point is if the API calls `setEnabled` then we should store the requested state.

If the API calls `setEnabled(false)` and then readonly is toggled on and off the state should still be `false`. If we haven't stored that, we might incorrectly enable it in `design` mode after that process.

Same with `setEnabled(true)`, no matter when the API is called that should be the "state" and we choose whether to render the state or override it based on `editor.readonly` and `allowedModes`.

Reply:
After further discussion this can be made a new ticket - a high priority one - for a 7.3.1 patch release.